### PR TITLE
doAfterNext operator like in RxJava

### PR DESCRIPTION
**Pull request by SlavinskiySergey**
Add doAfterNext operator like in RxJava

### CODE REVIEW

**Comment by kzaher** [comment]
Lets add this in 5.0. It will be safer to not break the ABI. Not huge issue, but let's wait a bit.

I think we would want to add entire corresponding `after` set and not just for `onNext`.

File: RxCocoa/Traits/SharedSequence/SharedSequence+Operators.swift [hunk 1]
```diff
@@ -108,16 +108,17 @@ extension SharedSequenceConvertibleType {
...
+     - parameter onAfterNext: Action to invoke for each element after the observable has passed an onNext event along to its downstream.
      - parameter onCompleted: Action to invoke upon graceful termination of the observable sequence.
      - parameter onSubscribe: Action to invoke before subscribing to source observable sequence.
      - parameter onSubscribed: Action to invoke after subscribing to source observable sequence.
      - parameter onDispose: Action to invoke after subscription to source observable has been disposed for any reason. It can be either because sequence terminates for some reason or observer subscription being disposed.
      - returns: The source sequence with the side-effecting behavior applied.
      */
-    public func `do`(onNext: ((E) -> Void)? = nil, onCompleted: (() -> Void)? = nil, onSubscribe: (() -> Void)? = nil, onSubscribed: (() -> Void)? = nil, onDispose: (() -> Void)? = nil)
+    public func `do`(onNext: ((E) -> Void)? = nil, onAfterNext: ((E) -> Void)? = nil, onCompleted: (() -> Void)? = nil, onSubscribe: (() -> Void)? = nil, onSubscribed: (() -> Void)? = nil, onDispose: (() -> Void)? = nil)
```
Comment (kzaher) [inline RxCocoa/Traits/SharedSequence/SharedSequence+Operators.swift]: Let's maybe call this just `afterNext` and add `afterCompleted` and `afterError` and add tests.

If you can make this change in less than one week, we can ship this in 5.0

File: RxSwift/Traits/Maybe.swift [hunk 2]
```diff
@@ -186,6 +187,7 @@ extension PrimitiveSequenceType where TraitType == MaybeTrait {
      - returns: The source sequence with the side-effecting behavior applied.
      */
     public func `do`(onNext: ((ElementType) throws -> Void)? = nil,
+                     onAfterNext: ((ElementType) throws -> Void)? = nil,
```
Comment (kzaher) [inline RxSwift/Traits/Maybe.swift]: The same rename here.

File: Tests/RxCocoaTests/SharedSequence+OperatorTest.swift [hunk 3]
```diff
@@ -283,6 +290,31 @@ extension SharedSequenceOperatorTests {
         let expectedEvents = [1, 2, -1]
         XCTAssertEqual(events, expectedEvents)
     }
+    
+    func testAsDriver_doOnAfterNext() {
+        let hotObservable = BackgroundThreadPrimitiveHotObservable<Int>()
+        
+        var events = [Int]()
```
Comment (kzaher) [inline Tests/RxCocoaTests/SharedSequence+OperatorTest.swift]: This doesn't test the ordering properly.

Please turn this into `var events = [OrderedEvent]()`

````
struct OrderedEvent {
    case before(Event<Int>)
    case subscription(Event<Int>)
    case after(Event<Int>)
}

hotObservable.asDriver(onErrorJustReturn: -1).do(onAfterNext: { e in
            XCTAssertTrue(DispatchQueue.isMain)
            events.append(.after(.next(e)))
        })
        .do(onNext: { e in
                 events.push(.subscribe(.next(e)))
        }
````

File: Tests/RxCocoaTests/SharedSequence+OperatorTest.swift [hunk 4]
```diff
@@ -217,6 +217,7 @@ extension SharedSequenceOperatorTests {
         let hotObservable = BackgroundThreadPrimitiveHotObservable<Int>()
 
         var events = [Event<Int>]()
+        var afterNextEvents = [Int]()
```
Comment (kzaher) [inline Tests/RxCocoaTests/SharedSequence+OperatorTest.swift]: This doesn't test the ordering properly.

Please turn this into `var events = [OrderedEvent]()`

````
struct OrderedEvent {
    case before(Event<Int>)
    case subscription(Event<Int>)
    case after(Event<Int>)
}

hotObservable.asDriver(onErrorJustReturn: -1).do(onAfterNext: { e in
            XCTAssertTrue(DispatchQueue.isMain)
            events.append(.after(.next(e)))
        })
        .do(onNext: { e in
                 events.push(.subscribe(.next(e)))
        }
```

in the end please assert that events are equal to some order.

File: RxCocoa/Traits/SharedSequence/SharedSequence+Operators.swift (hunk 1 above)
Comment (SlavinskiySergey) [reply RxCocoa/Traits/SharedSequence/SharedSequence+Operators.swift]: Done.

File: RxSwift/Traits/Maybe.swift (hunk 2 above)
Comment (SlavinskiySergey) [reply RxSwift/Traits/Maybe.swift]: Done.

File: Tests/RxCocoaTests/SharedSequence+OperatorTest.swift (hunk 3 above)
Comment (SlavinskiySergey) [reply Tests/RxCocoaTests/SharedSequence+OperatorTest.swift]: Done.

**Comment by SlavinskiySergey** [comment]
@kzaher thank your for advices. All your suggestions were applied.

**Comment by freak4pc** [comment]
@SlavinskiySergey Please run `./scripts/package-spm.swift`, and rebase your branch on top of develop. Thanks !

**Comment by kzaher** [comment]
@SlavinskiySergey I've noticed that you've just added it to `Maybe`. We would want to add
```swift
afterSuccess: ((ElementType) throws -> Void)? = nil,
afterError: ((Swift.Error) throws -> Void)? = nil,
```
to `Single` and 
```
afterError: ((Swift.Error) throws -> Void)? = nil,
afterCompleted: (() throws -> Void)? = nil,
```
to `Completable`.

Like @sergdort says
> Please run ./scripts/package-spm.swift, and rebase your branch on top of develop. Thanks !

**Comment by freak4pc** [comment]
I am a huge fan of @sergdort, but definite not the same person 😸

**Comment by SlavinskiySergey** [comment]
@freak4pc I run `./scripts/package-spm.swift` but nothing changed (git doesn't show any changes).

**Comment by freak4pc** [comment]
I ran it on your branch and definitely seeing changes. Can you make sure you have nothing stages on git?

**Comment by kzaher** [comment]
@freak4pc You all look like a bunch of alphanumeric characters to me, it's easily to get confused from where I'm standing :)

**Comment by freak4pc** [comment]
🤖

**Comment by kzaher** [comment]
Quickly, show me some captcha!

**Comment by freak4pc** [comment]
@SlavinskiySergey I've done the changes on this branch: https://github.com/ReactiveX/RxSwift/tree/doAfterNext

Feel free to cherry pick https://github.com/ReactiveX/RxSwift/commit/22e3e76c5ae631e1036f601a69aea287d0090958 if you want to.

If you have any issues I can open a PR with my branch - with your commits in it. 

Let me know if I can help!

**Comment by SlavinskiySergey** [comment]
@kzaher I pushed changes for `Single` and `Completable`.
@freak4pc thank you for your help.

**Comment by freak4pc** [comment]
You still don't have the updated tests file. Are you able to resolve the package-spm.swift thing ?

**Comment by SlavinskiySergey** [comment]
I see errors when run in console:

`error: could not build module 'Darwin'
#include <Availability.h>
         ^
<unknown>:0: error: could not build Objective-C module 'CoreFoundation'`

**Comment by freak4pc** [comment]
@SlavinskiySergey Do you have Xcode 10.2 selected?

In terminal if you type `xcode-select -p`, which path do you see?

**Comment by SlavinskiySergey** [comment]
@freak4pc Yes, I have Xcode10.2
 In terminal if type `xcode-select -p`  I see path to my Xcode (  `/Applications/Xcode.app/Contents/Developer` )

**Comment by freak4pc** [comment]
I see. Would it be OK If I make a new PR with your commits and perform the needed changes? @SlavinskiySergey

**Comment by SlavinskiySergey** [comment]
@freak4pc it will be OK. Thank's a lot!

**Comment by freak4pc** [comment]
Did this in #1926.

**Pull request closed without merging**